src/coreclr/vm/jitinterface.cpp

```diff
@@ -6058,8 +6061,7 @@ CorInfoHelpFunc CEEInfo::getCastingHelperStatic(TypeHandle clsHnd, bool fThrowin
             *pfClassMustBeRestored = true;
         }
 
-        // If it is an array, use the fast array helper
-        helper = CORINFO_HELP_ISINSTANCEOFARRAY;
+        _ASSERTE(helper == CORINFO_HELP_ISINSTANCEOFANY);
     }
     else
     if (!clsHnd.IsTypeDesc() && !Nullable::IsNullableType(clsHnd))
```

Comment:
The native AOT implementation does not checks for Nullable. Is this check redundant here or is the check for Nullable missing in native AOT?

Reply:
The difference is that JIT tries to transform simple cases like `o is int?` into `o is int`, but can't do that when needs a type lookup. As I understand that is due to limitations of IR.

https://github.com/dotnet/runtime/blob/40b39ff7df1dc2388a5865c9ad151dce88fa007d/src/coreclr/jit/importer.cpp#L5387-L5391

NativeAOT does not seem to have a problem to express such lookup, so we always cast with nullable stripped

https://github.com/dotnet/runtime/blob/17d13fa10be9020c851e8f2e668685b4cb98d6eb/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Compilation.cs#L301-L306

Reply:
In other words in NativeAOT the type check will always be to an unboxed underlying type, thus CLASS variant is correct, and also optimizable, since T? is `isClassExact`.

In CoreCLR the type check in complex cases could be to the actual `Nullable<SomeStruct<string>>`, thus it should use ANY variant. It also means that we may be introducing an optimization bug, since such IsInst cannot be lowered to a type/handle compare.

I will check if that is the case.

Reply:
Sadly that is the case. The following works incorrectly

```cs
===== Prints:

True
False


using System.Runtime.CompilerServices;

namespace ConsoleApp34
{
    struct S1<T>
    {
        public T value;
    }

    internal class Program
    {
        static object o;
        static void Main(string[] args)
        {
            Test<int>();
            Test<string>();
        }

        [MethodImpl(MethodImplOptions.AggressiveOptimization)]
        private static void Test<T>()
        {
            o = new S1<T>();

            Console.WriteLine(o is S1<T>?);
        }
    }
}
```

This PR indirectly enabled casting optimizations for cases like `o is int?`, but we need to suppress it, since in more general cases it does not work correctly.

Reply:
Unless there are better ideas, I am thinking of constraining isinst optimization for `CORINFO_HELP_ISINSTANCEOFANY` only if converting to an array type. So we do not keep finding more broken cases.

That may be too conservative, but we should probably stay closer to the preexisting behavior for now and consider if more cases can work in 9.0

Reply:
I am thinking that this PR as a whole is too risky for 8.0. Are there parts that we think are critical to get into .NET 8?

Reply:
I think the actual casting helper change was low risk as that is mostly refactoring to get more cases to hit the cache earlier. Touching the JIT appears to be a lot more fragile.

There is nothing really "critical" for 8.0, as in - we are not fixing some complete showstoppers here.

Reply:
I have created a "reduced" version of this. I think that is what we can consider for 8.0 - https://github.com/dotnet/runtime/pull/90234

I have removed all the JIT changes, but kept the added codegen test.

Reply:
Sounds good.